File: java-sdks/src/main/java/com/suisdks/sui/transactions/intents/CoinWithBalance.java
package com.suisdks.sui.transactions.intents;

import java.math.BigInteger;

public class CoinWithBalance {

    public static final String INTENT_NAME = "CoinWithBalance";
    private static final String SUI_TYPE = "0x2::sui::SUI";

    public static CoinWithBalanceIntentBuilder coinWithBalance(BigInteger balance, String type, boolean useGasCoin) {
        String coinType = "gas".equals(type) ? SUI_TYPE : type;
        return new CoinWithBalanceIntentBuilder(coinType, balance, useGasCoin);
    }

    public static CoinWithBalanceIntentBuilder coinWithBalance(BigInteger balance) {
        return coinWithBalance(balance, SUI_TYPE, true);
    }

    public static CoinWithBalanceIntentBuilder coinWithBalance(BigInteger balance, String type) {
        return coinWithBalance(balance, type, true);
    }

    public static class CoinWithBalanceIntentBuilder {

        private final String type;
        private final BigInteger balance;
        private final boolean useGasCoin;

        public CoinWithBalanceIntentBuilder(String type, BigInteger balance, boolean useGasCoin) {
            this.type = type;
            this.balance = balance;
            this.useGasCoin = useGasCoin;
        }

        public CoinWithBalanceIntent build() {
            return new CoinWithBalanceIntent(INTENT_NAME, type, balance, useGasCoin);
        }
    }

    public record CoinWithBalanceIntent(
            String name,
            String type,
            BigInteger balance,
            boolean useGasCoin
    ) {

        public static CoinWithBalanceIntent fromMap(java.util.Map<String, Object> data) {
            return new CoinWithBalanceIntent(
                    (String) data.get("name"),
                    (String) data.get("type"),
                    new BigInteger(data.get("balance").toString()),
                    (Boolean) data.getOrDefault("useGasCoin", true)
            );
        }

        public java.util.Map<String, Object> toMap() {
            java.util.Map<String, Object> map = new java.util.HashMap<>();
            map.put("name", name);
            map.put("type", type);
            map.put("balance", balance);
            map.put("useGasCoin", useGasCoin);
            return map;
        }
    }

    public static class CoinWithBalanceResolver {

        public java.util.Map<String, Object> resolveCoinBalance(
                java.util.List<CoinWithBalanceIntent> intents,
                String senderAddress
        ) {
            java.util.Set<String> coinTypes = new java.util.HashSet<>();
            java.util.Map<String, BigInteger> totalByType = new java.util.HashMap<>();

            for (CoinWithBalanceIntent intent : intents) {
                if (!"gas".equals(intent.type()) && intent.balance().compareTo(BigInteger.ZERO) > 0) {
                    coinTypes.add(intent.type());
                }
                totalByType.merge(intent.type(), intent.balance(), BigInteger::add);
            }

            java.util.Map<String, Object> result = new java.util.HashMap<>();
            result.put("coinTypes", coinTypes);
            result.put("totalByType", totalByType);
            result.put("sender", senderAddress);

            return result;
        }
    }
}

File: kotlin-sdks/src/main/kotlin/com/suisdks/sui/transactions/intents/CoinWithBalance.kt
# Intent system for Kotlin SDK
package com.suisdks.sui.transactions.intents

import java.math.BigInteger

const val COIN_WITH_BALANCE = "CoinWithBalance"
const val SUI_TYPE = "0x2::sui::SUI"

/**
 * CoinWithBalance represents the CoinWithBalance intent
 */
data class CoinWithBalance(
    val name: String,
    val type: String,
    val balance: BigInteger,
    val useGasCoin: Boolean
)

/**
 * CoinWithBalanceBuilder builds a CoinWithBalance intent
 */
class CoinWithBalanceBuilder(
    private val balance: BigInteger
) {
    private var coinType: String = SUI_TYPE
    private var useGasCoin: Boolean = true
    
    fun setCoinType(type: String): CoinWithBalanceBuilder {
        this.coinType = type
        return this
    }
    
    fun useGasCoin(use: Boolean): CoinWithBalanceBuilder {
        this.useGasCoin = use
        return this
    }
    
    fun build(): CoinWithBalance {
        return CoinWithBalance(
            name = COIN_WITH_BALANCE,
            type = coinType,
            balance = balance,
            useGasCoin = useGasCoin
        )
    }
}

/**
 * Creates a CoinWithBalance intent
 */
fun coinWithBalance(balance: BigInteger): CoinWithBalanceBuilder {
    return CoinWithBalanceBuilder(balance)
}

/**
 * Creates a CoinWithBalance intent with type
 */
fun coinWithBalanceType(balance: BigInteger, type: String): CoinWithBalanceBuilder {
    return CoinWithBalanceBuilder(balance).setCoinType(type)
}

/**
 * IntentResolver resolves intents
 */
interface IntentResolver {
    suspend fun resolve(intent: CoinWithBalance): kotlinx.coroutines.Deferred<Unit>
}

/**
 * CoinWithBalanceResolver resolves CoinWithBalance intents
 */
class CoinWithBalanceResolver(
    private val sender: String
) : IntentResolver {
    
    override suspend fun resolve(intent: CoinWithBalance): kotlinx.coroutines.Deferred<Unit> {
        return kotlinx.coroutines.async {
            // Resolve logic here
        }
    }
    
    /**
     * Resolves coin balance for CoinWithBalance intents
     */
    fun resolveCoinBalance(
        intents: List<CoinWithBalance>,
        sender: String
    ): Map<String, Any> {
        val coinTypes = mutableSetOf<String>()
        val totalByType = mutableMapOf<String, BigInteger>()
        
        for (intent in intents) {
            if (intent.type != "gas" && intent.balance > BigInteger.ZERO) {
                coinTypes.add(intent.type)
            }
            totalByType[intent.type] = (totalByType[intent.type] ?: BigInteger.ZERO) + intent.balance
        }
        
        return mapOf(
            "coinTypes" to coinTypes,
            "totalByType" to totalByType,
            "sender" to sender
        )
    }
}

/**
 * IntentSystem manages intents
 */
class IntentSystem {
    private val resolvers = mutableMapOf<String, IntentResolver>()
    
    /**
     * Adds an intent resolver
     */
    fun addResolver(name: String, resolver: IntentResolver) {
        resolvers[name] = resolver
    }
    
    /**
     * Resolves an intent
     */
    suspend fun resolve(intent: CoinWithBalance) {
        val resolver = resolvers[intent.name]
            ?: throw IllegalArgumentException("No resolver for intent: ${intent.name}")
        resolver.resolve(intent).await()
    }
}

/**
 * IntentScope represents the scope of an intent
 */
enum class IntentScope(val value: Int) {
    TRANSACTION_DATA(0),
    PERSONAL_MESSAGE(1),
    TRANSACTION(3)
}